apps/cmp.c

```diff
@@ -0,0 +1,3557 @@
+        goto die;
+    }
+    /* See if an empty password will do */
+    if (PKCS12_verify_mac(p12, "", 0) || PKCS12_verify_mac(p12, NULL, 0)) {
+        pass = "";
+    } else {
+        if (pem_cb == NULL)
+            pem_cb = wrap_password_callback;
+        len = pem_cb(tpass, PEM_BUFSIZE, 0, cb_data);
+        if (len < 0) {
+            BIO_printf(bio_err, "passphrase callback error for %s\n", desc);
+            goto die;
+        }
+        if (len < PEM_BUFSIZE)
+            tpass[len] = 0;
+        if (!PKCS12_verify_mac(p12, tpass, len)) {
+            BIO_printf(bio_err,
+                       "mac verify error (wrong password?) in PKCS12 file for %s\n",
+                       desc);
+            goto die;
+        }
+        pass = tpass;
+    }
+    ret = PKCS12_parse(p12, pass, pkey, cert, ca);
+    if (ret && *ca != NULL) {
+        int i; /* other certs are for some reason in reverted order */
+        STACK_OF(X509) *certs = sk_X509_new_null();
+        for (i = 0; i < sk_X509_num(*ca); i++)
+            if (certs == NULL
+                    || !sk_X509_insert(certs, sk_X509_value(*ca, i), 0)) {
+                sk_X509_pop_free(certs, X509_free);
+                sk_X509_pop_free(*ca, X509_free);
+                X509_free(*cert);
+                EVP_PKEY_free(*pkey);
+                ret = 0;
+                goto die;
+            }
+        sk_X509_free(*ca);
+        *ca = certs;
+    }
+
+ die:
+    PKCS12_free(p12);
+    return ret;
+}
+
+/*
+ * TODO remove when load_cert_pass() is merged in apps/lib/apps.c (PR #4930)
+ * and after generalizing it w.r.t. timeout
+ */
+static X509 *load_cert_pass(const char *file, int format, const char *pass,
+                            const char *cert_descrip)
+{
+    X509 *x = NULL;
+    BIO *cert = NULL;
+    PW_CB_DATA cb_data;
+
+    cb_data.password = pass;
+    cb_data.prompt_info = file;
+
+    if (format == FORMAT_HTTP) {
+#if !defined(OPENSSL_NO_SOCK)
+        x = X509_load_http(file, NULL, NULL, 10);
+#else
+        CMP_err1("http(s) not supported loading cert from '%s'\n", file);
+#endif
+        goto end;
+    }
+    if (file == NULL) {
+        unbuffer(stdin);
+        cert = dup_bio_in(format);
+    } else {
+        cert = bio_open_default(file, 'r', format);
+    }
+    if (cert == NULL)
+        goto end;
+    if (format == FORMAT_ASN1) {
+        x = d2i_X509_bio(cert, NULL);
+    } else if (format == FORMAT_PEM) {
+        x = PEM_read_bio_X509_AUX(cert, NULL, wrap_password_callback, &cb_data);
+    } else if (format == FORMAT_PKCS12) {
+        EVP_PKEY *pkey = NULL; /* pkey is required for matching the cert */
+
+        if (!load_pkcs12(cert, cert_descrip, wrap_password_callback,
+                         &cb_data, &pkey, &x, NULL))
+            goto end;
+        EVP_PKEY_free(pkey);
+    } else {
+        BIO_printf(bio_err, "bad input format specified for %s file '%s'\n",
+                   cert_descrip, file);
+        goto end;
```

Comment:
"goto end" not needed?

Reply:
Yes. I've removed it and did some further minor code sanitzation to `load_cert_pass()`.